src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +23,78 @@
 class ExceptionSink(object):
   """A mutable singleton object representing where exceptions should be logged to."""
 
-  _destination = os.getcwd()
+  # TODO: see the bottom of this file where we call set_destination() and friends in order to
+  # properly setup global state.
+  # TODO: ???
+  _destination = None
+  # We need an exiter in order to know what to do after we log a fatal exception.
+  _exiter = None
+  # Where to log stacktraces to in a SIGUSR2 handler.
+  _trace_stream = None
+
+  _all_destinations = []
 
   def __new__(cls, *args, **kwargs):
     raise TypeError('Instances of {} are not allowed to be constructed!'
                     .format(cls.__name__))
 
+  class ExceptionSinkError(Exception): pass
+
+  # TODO (in response to the above): we can avoid a single entry point and make everything work with
+  # great ease if we can ensure all set_* methods are idempotent!
   @classmethod
   def set_destination(cls, dir_path):
-    cls._destination = dir_path
+    cls._destination = cls._check_or_create_new_destination(dir_path)
+    cls._all_destinations.append(cls._destination)
 
   @classmethod
   def get_destination(cls):
     return cls._destination
 
   @classmethod
-  def exceptions_log_path(cls, for_pid=None):
+  def exceptions_log_path(cls, for_pid=None, in_dir=None):
+    if not in_dir:
+      in_dir = cls.get_destination()
     intermediate_filename_component = '.{}'.format(for_pid) if for_pid else ''
     return os.path.join(
-      cls.get_destination(),
+      in_dir,
       'logs',
       'exceptions{}.log'.format(intermediate_filename_component))
 
+  @classmethod
+  def set_exiter(cls, exiter):
+    assert(isinstance(exiter, Exiter))
+    cls._exiter = exiter
+    sys.excepthook = cls._log_unhandled_exception_and_exit
+
+  @classmethod
+  def set_trace_stream(cls, trace_stream):
+    # TODO: validate trace_stream somehow!
+    # TODO: do we need to keep a reference to this? I think not.
+    cls._trace_stream = trace_stream
+
+    # TODO: move /all/ signal handling into a SignalHandler class defined in this file which is not
+    # a singleton and can be subclassed (like Exiter). Since signals are global, it is useful to
+    # be able to look at all the signal handlers which are set in one place.
+    if faulthandler.is_enabled():
+      faulthandler.disable()
+    faulthandler.enable(trace_stream)
+    # This permits a non-fatal `kill -31 <pants pid>` for stacktrace retrieval.
+    faulthandler.register(signal.SIGUSR2, trace_stream, chain=True)
+
+  @classmethod
+  def _check_or_create_new_destination(cls, destination):
+    try:
+      safe_mkdir(destination)
+    except Exception as e:
+      # NB: When this class sets up excepthooks, raising this should be safe, because we always
+      # have a destination to log to (os.getcwd() if not otherwise set).
```

Comment:
> Now that this field defaults to None this comment isn't accurate.

Will update.

> But regardless: I think that if this is refactored to send the signal-handler output to the exception log, it won't matter, because we'll need to eagerly open the exception/signal log in order to register it with faulthandler.

I'm understanding this as: we need to handle "opening the exception log in a signal/exception handler" anyway so in doing so we don't have to worry about the validity of the exception log in other methods (because it'd be handled specifically somewhere).

Reply:
> "opening the exception log in a signal/exception handler" anyway

See the faulthandler docs linked above.